### ISSUE TITLE
Graph Positional and Structural Encoder

### ISSUE THREAD

**Issue by rusty1s** [pyg-team/pytorch_geometric#8310]
### 🚀 The feature, motivation and pitch

It would be great to add the "GPSE" (Graph Positional and Structural Encoder) module to PyG, which acts as a trained encoder for Graph Transformers. Most importantly, we should support loading pre-trained encoders and provide them to the user, e.g., via
```python
encoder = GPSE.from_pretrained("molpcba")
```

Code: https://github.com/G-Taxonomy-Workgroup/GPSE
Paper: https://arxiv.org/abs/2307.07107
Twitter: https://twitter.com/RemyLau3/status/1717194030986190899

Any help on this is highly appreciated :)

**Comment by MrWeast** [comment]
I would like to get involved and to be assigned this issue.

**Comment by MrWeast** [comment]
can you provide more details on what/how you want things incorporated? Do you just want a way to load the pretrained models from the provided code? 

Also if there is someone active in the community I could collaborate with on this I would appreciate it! I would like to quickly become more familiar with how things are done in the PyG community so I can work on other issues as well!

**Comment by rusty1s** [comment]
Yes, I don't think we should necessarily provide the full code to train GPSE from scratch. For me, it would be sufficient if people have an easy time to try out pre-trained `GPSE` on their own from PyG. Happy to help and discuss any questions.

**Comment by MrWeast** [comment]
sorry it has taken awhile for me to get back to this I've been really busy. 

Two things:

1. --------------------------------------------------
I'm trying to see how they are loading a pretrained model in the provided code and starting on line 136 of their main.py file they have:

if cfg.pretrained.dir:
          cfg = load_pretrained_model_cfg(cfg)

however, in none of their provided config files have the key pretrained and I can't seem to find where they might be creating it.

do you have any advice on this?
2. --------------------------------------------------
Would you prefer the pre-trained parameters be already downloaded for the user or have them downloaded via wget or some other method at runtime?

**Comment by rusty1s** [comment]
1. This seems to happen here (https://github.com/G-Taxonomy-Workgroup/GPSE/blob/main/main.py#L146) and here (https://github.com/G-Taxonomy-Workgroup/GPSE/blob/main/graphgym/finetuning.py#L99).
2. At best, parameters are downloaded on-the-fly, similar to how `SchNet.from_pretrained` works or any of our datasets.

Still super thankful you take a look at this :)

**Comment by MrWeast** [comment]
I seem to be having a hard time with this, I can't even get a pretrained model to load following their own instructions on their code. I get errors related to loading the state dictionary to a model with improper dimensions.

I downloaded a pretrained model via:

# Pre-trained on MolPCBA (default)
wget https://zenodo.org/record/8145095/files/gpse_model_molpcba_1.0.pt -O pretrained_models/gpse_molpcba.pt

and then try to run downstream evaluations via:
python main.py --cfg configs/mol_bench/zinc-GPS+GPSE.yaml

and then I get a long list of errors like this: 
RuntimeError: Error(s) in loading state_dict for GraphGymModule:
        size mismatch for model.post_mp.node_post_mps.0.model.0.Layer_0.layer.model.weight: copying a param with shape torch.Size([32, 512]) from checkpoint, the shape in current model is torch.Size([512, 512]).

do you have any advice for this?

**Comment by rusty1s** [comment]
I would need to take a look. Can we create an issue for this on their GitHub?

**Comment by MrWeast** [comment]
sure thing, I will do that now

**Comment by MrWeast** [comment]
I think the issue might be pyg-lib, I'm running on windows 10 and I can see in the documentation that windows is not supported but I didn't get any errors when installing their dependencies to run their code (of which pyg-lib is one)

**Comment by rusty1s** [comment]
Interesting. How does the shape mismatch relate to `pyg-lib`? I am also not sure if the GPSE project uses `pyg-lib` at all (since it is graph classification).

**Comment by MrWeast** [comment]
I just assumed since it was a dependency and incompatible with my environment. But apparently I'm also using the wrong pyg version for their code so it could be that as well.

**Comment by brovatten** [comment]
Are you still working on this @MrWeast ? If not I could also give this a try!

**Comment by MrWeast** [comment]
@brotvatten be my guest. I've had to step away from working on this for a bit since I'm in my final exam week and trying to get a paper published. I was planning on getting back to it when I can but I certainly don't want to slow down development.

**Comment by brovatten** [comment]
Okay, I'll take a look at it at least. Good luck with your exams :)

**Comment by semihcanturk** [comment]
Hi @rusty1s, @MrWeast & @brovatten! I'm Semih, a main author of GPSE alongside @RemyLau. Many thanks for your interest in integrating GPSE, and we'd be happy to help with the integration and have our work as part of PyG.

For starters, I realized that @MrWeast [has already reached out to us regarding the aforementioned GPSE issue](https://github.com/G-Taxonomy-Workgroup/GPSE/issues/1), which was caused by a bug in PyG 2.4 regarding the GraphGym MLP LayerConfig. The related PyG issue & hotfix can be found here: https://github.com/pyg-team/pytorch_geometric/issues/8484. In short, you should be able to use GPSE with PyG 2.4 with no issues now.

Is either @MrWeast / @brovatten currently working on this? We wouldn't want to override anyone else's work, but are more than happy to assist/collaborate or just make a PR with the desired functionality. Please let me know :)

**Comment by MrWeast** [comment]
sorry I've actually been meaning to un-assign myself, I think I have too much on my plate currently to focus on this. I will try to come back to get involved/contribute in the future. @brovatten or @semihcanturk you are welcome to take over

**Comment by semihcanturk** [comment]
Hey @rusty1s, I've made a PR above for GPSE. There are few minor things I'll handle the next day or two (mostly for prettifying the example), but it's ready for review now. Let me know if you have any questions, and any suggestions/edits are of course welcome.

**Comment by rusty1s** [comment]
Thank you! Since reviewing the full PR is hard, can we split the model implementation from the example and craft two PRs?

**Comment by semihcanturk** [comment]
Yeah sure! How do we want to go about this? Should I just close the previous PR and open two new ones?

**Comment by rusty1s** [comment]
Let's hold on for now and get this in first, thank you! :)
